job-shop-scheduling/job_shop_scheduler.py

```diff
@@ -285,6 +297,8 @@ def get_bqm(self):
         # - Therefore, with this penalty scheme, all optimal solution penalties < any non-optimal
         #   solution penalties
         base = len(self.last_task_indices) + 1     # Base for exponent
+        # Get our pruned (remove_absurd_times) variable list so we don't undo pruning
+        pruned_variables = list(bqm.variables)
```

Comment:
I don't understand the purpose of this. Could you please elaborate?

`list(bqm.variables)` does not contain the variables that there pruned in `remove_absurd_times`. This is because once we fix the variable to a particular value (as done in `remove_absurd_times` with the `fix_variable(..)` method)

You can see test this with the following test:
```
import dimod
bqm = dimod.BinaryQuadraticModel({'a':2, 'b':1}, {'ab':1}, 0, dimod.SPIN)
print(bqm)

# Now fix/prune the variable 'b'
bqm.fix_variable('b', 1)
print(bqm)
```

Reply:
Sorry, this was poor variable naming. Instead of meaning "variables that were pruned out of the csp" I meant "csp variables post pruning". 

Since the code after this sets biases for each time variable as long as the time + duration does not exceed the JSP max_time, we may re-introduce a variable that was removed in the pruning phase. The list is checked to make sure that we only edit variables that already exist in the BQM.

Do you have any ideas for a better way to do this, and for a better variable name?

Reply:
I see the issue you've identified - you're right, the pruning is being un-done.

The code as it is written right now, would work, but it could be more efficient. Namely, the line `if label in pruned_variables:` is expensive when it is found within two for-loops. The if-x-is-in-my_list code is O(n), but putting it within those for-loops, makes that check particularly expensive.

I would suggest a more invasive change to the code. Namely, the new `get_bqm(..)` could:
* Apply the 3 constraints: `self._add_one_start_constraint()`, `self._add_precedence_constraint()`, `self._add_share_machine_constraint()`
* Apply the the biases to encourage shortest schedule, perhaps with a new function called `self._add_biases_for_shortest_schedule()`
* Finally, apply pruning last. (i.e. apply self._remove_absurd_times())

By having the pruning last, we won't have to work as hard to keep track of our variables.

Reply:
To be more explicit,
```
def get_bqm(..):
    if stitch_kwargs == None:
        ..

    # Apply constraints to self.csp
    self._add_one_start_constraint()
    self._add_precedence_constraint()
    self._add_share_machine_constraint()

    # Get BQM
    self.bqm = dwavebinarycsp.stitch(self.csp, **stitch_kwargs)

    # Edit BQM to encourage shortest schedule and to prune variables
    self._add_biases_for_shortest_schedule()
    self._remove_absurd_times()

    return self.bqm
```

Changes:
* should probably add `self.bqm = None` to `self.__init__(..)`
* note that in my suggestion, `_remove_absurd_times` is occurring AFTER `stitch(..)`, so it will be dealing with the BQM directly. Hence, `self._remove_absurd_times()` would be working with `self.bqm` rather than `self.csp` (i.e. `self.bqm.fix_variable(..)`). See `BinaryQuadraticModel.fix_variable(..)`  [documentation](https://docs.ocean.dwavesys.com/projects/dimod/en/latest/reference/binary_quadratic_model.html#transformations)

Sorry for the really long message. Thank you very much for noticing this error in the demo and for taking the initiative to fix it!

Reply:
Now would probably be a good time to mention another issue that I had. To get the CSP, I have to call get_bqm (or the CSP constraint and pruning functions directly, but there is another issue with that). This seems out of place to me and I would personally prefer the CSP functions to be separate (first thought is in process_data).
Supposing I do call the CSP functions directly to get just the CSP without creating the BQM, if (more likely when) I call get_bqm, the CSP constraints will be unnecessarily duplicated. Similarly, if I instead wanted to try a different stitch_kwargs, the constraints would again be added.
Also, if the BQM variables are fixed (instead of the CSP), it looks like csp.check will throw a key error when checking the sample for the pruned BQM. I'm assuming samples for that BQM will be missing variables that the CSP has. While the same variables can simply be pruned in the CSP, something different might need to be done if the above points warrant changes.
Thoughts?

Reply:
On CSP:
* Interesting. When writing this code, I only imagine the users importing the function `get_jss_bqm(..)` (which is not part of the `JobShopScheduler` class). The class `JobShopScheduler` was merely a convenient way for me to modularize the code and generate a JSS bqm (read: I did not intend for the user to interact with `JobShopScheduler` at all).
* Could clarify by what you mean by "first thought is in `process_data(..)`"? At the moment, I am interpreting it as wanting to put the CSP functions into `_process_data(..)`, which can't be right as `_process_data(..)` is processing the user's JSS dict.
* I didn't intend for the user to interact with `JobShopScheduler`, but should they wish to, your proposed function seems reasonable. Would you be willing to add that in?

Reply:
On BQM variables fixed:
* Could you please send me the error you are seeing for this - "it looks like csp.check will throw a key error when checking the sample for the pruned BQM"? I would also expect that the CSP would contain variables that the BQM has pruned, so I'm not sure why the `csp.check` would throw a key error (i.e. I see CSP containing the superset of variables and hence, did not expect this error)
* As for maintaining a pruned CSP in a world where we have `get_csp(..)` in JSS. I'm not sure how useful this pruned CSP would be. The effort made to prune the CSP variables initially was so that the *BQM* is simpler when it gets fed to the QPU. Now the pruning process is just happening directly on the BQM. Thoughts?

Reply:
CSP:

- In that case it might just be best to leave it as is.
- I actually did mean within _process_data(), probably because when I think of a JSP, I just associate it with a CSP. I'm not really sure where you think it would be best to move it (if at all), I just wanted to mention that I felt it could be improved based on my experience.
- Being that you didn't intend for users to interact with JobShopScheduler directly, I'd prefer to leave it as is. The benefits (access to CSP without having to create the BQM, i.e. convenience) don't really seem worth the changes, I'll try to explain more below in one of the BQM points. You've made some really good suggestions, maybe you can think of something I haven't.

BQM:

- Sure, and here is the code I ran to produce the error. Btw the code is pretty much the same as in test_dense_schedule, so we need to prune the CSP for the tests to pass.
```
from job_shop_scheduler import JobShopScheduler
from tabu import TabuSampler       
jobs = {"j0": [(1, 2), (2, 2), (3, 2)],
        "j1": [(3, 4), (2, 1), (1, 1)],
        "j2": [(2, 2), (1, 3), (2, 1)]}
max_time = 7
jsp = JobShopScheduler(jobs, max_time)
bqm = jsp.get_bqm()
csp = jsp.csp
print(list(bqm.variables))
bqm.fix_variable(list(bqm.variables)[0], 1)
print(list(bqm.variables))
response = TabuSampler().sample(bqm, num_reads=2000)
response_sample, sample_energy, _ = next(response.data())
# the check done in test_dense_schedule
csp.check(response_sample)
```
Error is:

> Traceback (most recent call last):
>   File "<input>", line 16, in <module>
>   File "C:\Users\Mike\Anaconda3\envs\dwave_demos\lib\site-packages\dwavebinarycsp\core\csp.py", line 194, in check
>     return all(constraint.check(solution) for constraint in self.constraints)
>   File "C:\Users\Mike\Anaconda3\envs\dwave_demos\lib\site-packages\dwavebinarycsp\core\csp.py", line 194, in <genexpr>
>     return all(constraint.check(solution) for constraint in self.constraints)
>   File "C:\Users\Mike\Anaconda3\envs\dwave_demos\lib\site-packages\dwavebinarycsp\core\constraint.py", line 356, in check
>     return self.func(*(solution[v] for v in self.variables))
>   File "C:\Users\Mike\Anaconda3\envs\dwave_demos\lib\site-packages\dwavebinarycsp\core\constraint.py", line 356, in <genexpr>
>     return self.func(*(solution[v] for v in self.variables))
> KeyError: 'j0_0,0'

Note:  If the code was kept as is, we could simply prune the variable from the CSP and BQM at the same time.
- So assuming we had a get_csp function that was separate from get_bqm, and the user was accessing JobShopScheduler directly, we would want to prune the CSP. The user might want to see which variables were pruned and which remain, or check some configuration to see if the CSP is satisfied or not. If get_bqm was then called, the BQM would initially have the same variables as the CSP, but when the biases are added to encourage shorter schedules, variables removed during the pruning phase would be added again. So either (maybe you can think of a better way?) variables would have to be pruned a second time on the BQM, or we would have change the code that adds bias to not "blindly" edit biases, but to instead determine if the variable has been pruned or not.

To summarize: I don't think it's worth it (based on your original intention of only get_jss_bqm being called, and minimal benefit from changing) to separate the CSP function from get_bqm.

Reply:
I just went to implement your proposed changes assuming we decided to keep the CSP function in get_bqm. I.e.
```
def get_bqm(..):
    if stitch_kwargs == None:
        ..

    # Apply constraints to self.csp
    self._add_one_start_constraint()
    self._add_precedence_constraint()
    self._add_share_machine_constraint()

    # Get BQM
    self.bqm = dwavebinarycsp.stitch(self.csp, **stitch_kwargs)

    # Edit BQM to encourage shortest schedule and to prune variables
    self._add_biases_for_shortest_schedule()
    self._remove_absurd_times()

    return self.bqm
```
I believe I understood you correctly, and if so I think we missed one critical point that becomes quite clear when making the above changes. Since the CSP is not pruned before stitch is called, the resulting BQM is more complex. In the case from my above comment (test_dense_schedule), max_graph_size actually has to be set to 9 to get the BQM, and the resulting BQM has more aux variables (18 now vs 0 before the changes) .

With this in mind I would say the CSP should be pruned prior to stitching. Do you agree?

If so, how do you think would be best to prune the CSP and BQM separately? Two nearly identical functions, except one for CSP and the other for BQM? One function with a parameter that selects CSP or BQM? 

Thanks, and sorry for the delay in  getting to this.

Reply:
On BQM error:
- Your code looks great for the most part. It's just missing the line where you fill the pruned variables with zeros. So I would recommend importing `fill_with_zeros(..)` and you should find the correct answer. Specifically,
```
from test_job_shop_scheduler import fill_with_zeros
from job_shop_scheduler import JobShopScheduler
..
response_sample, sample_energy, _ = next(response.data()
fill_with_zeros(response_sample, jobs, max_time)    # fills the pruned values with 0
csp.check(response_sample) 
```

Reply:
You're right. Having a complicated CSP could lead to more auxiliary variables in our final BQM. Perhaps we could store the time range of each task during `_remove_absurd_times(..)` and then call up this time range when we add biases to encourage the shortest schedule? That way, we won't be un-doing the pruning from earlier while adding the biases to encourage the shortest schedule.

Reply:
Also, please don't apologize about the delay. I really appreciate you taking the time to help me improve this demo.

Reply:
I've looked through the code. I think overall, the existing changes improve the demo, so I would like to merge.

The following code issues that we've identified can either be addressed in a future PR or are simply a "nice to have, but not necessary":
* `_get_csp(..)` -> nice to have, but not essential
* `if label in pruned_variables` within multiple for-loops -> this can be addressed in a later PR
* pruning CSP and BQM -> the current method of pruning in the CSP and then avoid undoing that work is an improvement to the existing JSS demo

I would prefer to address our other issues in another PR as I think the scope of this PR is getting too large. Please let me know if you're okay with that and I will merge the existing code. (If you aren't okay with that, please let me know :P)

Reply:
I'm fine with merging now and leaving the issues you mentioned to separate PRs. I'll finish up the remaining issue (adding a good_stitch_kwargs for testing) and I think that will be everything. 

Let me know if I missed anything.

Reply:
That sounds great! Thank you for your contribution to this code base! :)